File: swift/Sources/FieldUICore/Engine/FieldStore.swift
#if canImport(simd)
import simd
#endif

/// FieldStore — owns the particle pool and the spatial index (§20.1 foundation).
///
/// The single home for "all the matter": add/remove particles, rebuild the
/// neighbour index once per frame, and answer `neighbors(_:r:)`. Count is the
/// conserved quantity (§2.4) — `size` is the live total.
public final class FieldStore {
    public private(set) var particles: [Particle] = []
    private let hash: SpatialHash

    public init(cellSize: Float = 64) {
        self.hash = SpatialHash(cellSize: cellSize)
    }

    public var size: Int { particles.count }

    @discardableResult
    public func add(_ p: Particle) -> Particle {
        particles.append(p)
        return p
    }

    /// Swap-remove a particle (O(1) after the scan, order not preserved).
    public func remove(_ p: Particle) {
        guard let i = particles.firstIndex(where: { $0 === p }) else { return }
        let last = particles.removeLast()
        if i < particles.count { particles[i] = last }
    }

    public func clear() {
        particles.removeAll()
        hash.clear()
    }

    /// Rebuild the neighbour index from the current pool (once per frame).
    public func reindex() {
        hash.rebuild(particles)
    }

    /// Neighbours within `r` of `p`, excluding `p` itself.
    public func neighbors(_ p: Particle, r: Float) -> [Particle] {
        hash.near(p.position, r: r).filter { $0 !== p }
    }

    /// Particles within `r` of an arbitrary point — for grid sampling, where the
    /// query origin is not itself a particle.
    public func near(_ point: Vec3, r: Float) -> [Particle] {
        hash.near(point, r: r)
    }
}

// MARK: - Energy diagnostics

/// Snapshot of kinetic, thermal, and total energy for the current pool.
/// The Swift port of `energyReport` from @field-ui/core/diagnostics/energy.
public func energyReport(_ particles: [Particle]) -> EnergyReport {
    var kinetic: Float = 0
    var thermal: Float = 0
    for p in particles {
        kinetic += 0.5 * p.mass * simd_length_squared(p.velocity)
        thermal += p.heat
    }
    return EnergyReport(kinetic: kinetic, thermal: thermal, total: kinetic + thermal, count: particles.count)
}

File: android/fundamental-core/src/main/kotlin/com/fundamental/core/engine/FieldStore.kt
package com.fundamental.core.engine

import com.fundamental.core.math.Vec3

/**
 * FieldStore — owns the particle pool and the spatial index (§20.1). The single home for "all the
 * matter": add/remove particles, rebuild the neighbour index once per frame, and answer
 * `neighbors(p, r)`. Count is the conserved quantity (§2.4); `size` is the live total.
 * Kotlin port of swift/Sources/FundamentalCore/Engine/FieldStore.swift.
 */
class FieldStore(cellSize: Float = 64f) {
    private val _particles = ArrayList<Particle>()
    val particles: List<Particle> get() = _particles
    private val hash = SpatialHash(cellSize)

    val size: Int get() = _particles.size

    fun add(p: Particle): Particle {
        _particles.add(p)
        return p
    }

    /** Swap-remove a particle (O(1) after the scan, order not preserved). */
    fun remove(p: Particle) {
        val i = _particles.indexOfFirst { it === p }
        if (i < 0) return
        val last = _particles.removeAt(_particles.size - 1)
        if (i < _particles.size) _particles[i] = last
    }

    fun clear() {
        _particles.clear()
        hash.clear()
    }

    /** Rebuild the neighbour index from the current pool (once per frame). */
    fun reindex() = hash.rebuild(_particles)

    /** Neighbours within [r] of [p], excluding [p] itself. */
    fun neighbors(p: Particle, r: Float): List<Particle> =
        hash.near(p.position, r).filter { it !== p }

    /** Particles within [r] of an arbitrary point (grid sampling, where the origin isn't a particle). */
    fun near(point: Vec3, r: Float): List<Particle> = hash.near(point, r)
}

/** Snapshot of kinetic, thermal, and total energy for the current pool. */
data class EnergyReport(val kinetic: Float, val thermal: Float, val total: Float, val count: Int)

/** The Kotlin port of `energyReport` (core/diagnostics/energy). */
fun energyReport(particles: List<Particle>): EnergyReport {
    var kinetic = 0f
    var thermal = 0f
    for (p in particles) {
        kinetic += 0.5f * p.mass * p.velocity.lengthSquared()
        thermal += p.heat
    }
    return EnergyReport(kinetic, thermal, kinetic + thermal, particles.size)
}
